File: example-source-500px/src/main/java/com/example/muzei/examplesource500px/FiveHundredPxExampleArtSource.java
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.muzei.examplesource500px;

import android.content.Intent;
import android.net.Uri;
import android.text.TextUtils;
import android.util.Log;

import com.google.android.apps.muzei.api.Artwork;
import com.google.android.apps.muzei.api.RemoteMuzeiArtSource;

import java.io.IOException;
import java.util.Random;

import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import static com.example.muzei.examplesource500px.FiveHundredPxService.Photo;
import static com.example.muzei.examplesource500px.FiveHundredPxService.PhotosResponse;

public class FiveHundredPxExampleArtSource extends RemoteMuzeiArtSource {
    private static final String TAG = "500pxExample";
    private static final String SOURCE_NAME = "FiveHundredPxExampleArtSource";

    private static final int ROTATE_TIME_MILLIS = 3 * 60 * 60 * 1000; // rotate every 3 hours

    public FiveHundredPxExampleArtSource() {
        super(SOURCE_NAME);
    }

    @Override
    public void onCreate() {
        super.onCreate();
        setUserCommands(BUILTIN_COMMAND_ID_NEXT_ARTWORK);
    }

    @Override
    protected void onTryUpdate(int reason) throws RetryException {
        String currentToken = (getCurrentArtwork() != null) ? getCurrentArtwork().getToken() : null;

        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .addInterceptor(new Interceptor() {
                    @Override
                    public Response intercept(final Chain chain) throws IOException {
                        Request request = chain.request();
                        HttpUrl url = request.url().newBuilder()
                                .addQueryParameter("consumer_key", Config.CONSUMER_KEY).build();
                        request = request.newBuilder().url(url).build();
                        return chain.proceed(request);
                    }
                })
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://api.500px.com/")
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        FiveHundredPxService service = retrofit.create(FiveHundredPxService.class);
        PhotosResponse response;
        try {
            response = service.getPopularPhotos().execute().body();
        } catch (IOException e) {
            Log.w(TAG, "Error reading 500px response", e);
            throw new RetryException();
        }

        if (response == null || response.photos == null) {
            throw new RetryException();
        }

        if (response.photos.size() == 0) {
            Log.w(TAG, "No photos returned from API.");
            scheduleUpdate(System.currentTimeMillis() + ROTATE_TIME_MILLIS);
            return;
        }

        Random random = new Random();
        Photo photo;
        String token;
        while (true) {
            photo = response.photos.get(random.nextInt(response.photos.size()));
            token = Integer.toString(photo.id);
            if (response.photos.size() <= 1 || !TextUtils.equals(token, currentToken)) {
                break;
            }
        }

        publishArtwork(new Artwork.Builder()
                .title(photo.name)
                .byline(photo.user.fullname)
                .imageUri(Uri.parse(photo.image_url))
                .token(token)
                .viewIntent(new Intent(Intent.ACTION_VIEW,
                        Uri.parse("http://500px.com/photo/" + photo.id)))
                .build());

        scheduleUpdate(System.currentTimeMillis() + ROTATE_TIME_MILLIS);
    }
}

File: example-source-500px/src/main/java/com/example/muzei/examplesource500px/FiveHundredPxExampleArtSource.kt
/*
 * Copyright 2014 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.example.muzei.examplesource500px

import android.content.Intent
import android.util.Log
import androidx.net.toUri
import com.example.muzei.examplesource500px.FiveHundredPxService.Photo
import com.google.android.apps.muzei.api.Artwork
import com.google.android.apps.muzei.api.RemoteMuzeiArtSource
import java.io.IOException
import java.util.Random

class FiveHundredPxExampleArtSource : RemoteMuzeiArtSource(SOURCE_NAME) {

    companion object {
        private const val TAG = "500pxExample"
        private const val SOURCE_NAME = "FiveHundredPxExampleArtSource"
        private const val ROTATE_TIME_MILLIS = 3 * 60 * 60 * 1000 // rotate every 3 hours
    }

    override fun onCreate() {
        super.onCreate()
        setUserCommands(BUILTIN_COMMAND_ID_NEXT_ARTWORK)
    }

    @Throws(RemoteMuzeiArtSource.RetryException::class)
    override fun onTryUpdate(@UpdateReason reason: Int) {
        val currentToken = currentArtwork?.token

        val photos = try {
            FiveHundredPxService.popularPhotos()
        } catch (e: IOException) {
            Log.w(TAG, "Error reading 500px response", e)
            throw RemoteMuzeiArtSource.RetryException()
        }

        if (photos.isEmpty()) {
            Log.w(TAG, "No photos returned from API.")
            scheduleUpdate(System.currentTimeMillis() + ROTATE_TIME_MILLIS)
            return
        }

        val random = Random()
        var photo: Photo
        var token: String
        while (true) {
            photo = photos[random.nextInt(photos.size)]
            token = Integer.toString(photo.id)
            if (photos.size <= 1 || token != currentToken) {
                break
            }
        }

        if (BuildConfig.DEBUG) {
            Log.d(TAG, "Loaded ${photo.id} with uri: ${photo.images[0].https_url}")
        }
        publishArtwork(Artwork.Builder()
                .title(photo.name)
                .byline(photo.user.fullname)
                .imageUri(photo.images[0].https_url?.toUri())
                .token(token)
                .viewIntent(Intent(Intent.ACTION_VIEW, "http://500px.com/photo/${photo.id}".toUri()))
                .build())

        scheduleUpdate(System.currentTimeMillis() + ROTATE_TIME_MILLIS)
    }
}

